monitoring/harmony-monitor/blockchain-watchdog/crosslink-monitor.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+  "encoding/json"
+  "fmt"
+  "strconv"
+  "sync"
+  "time"
+)
+
+// Only need to query leader on Shard 0
+func (m *monitor) crossLinkMonitor(interval uint64, poolSize int, pdServiceKey, chain string, shardMap map[string]int) {
+  crossLinkRequestFields := getRPCRequest(LastCrossLinkRPC)
+  nodeRequestFields := getRPCRequest(NodeMetadataRPC)
+
+  jobs := make(chan work, len(shardMap))
+  replyChannels := make(map[string](chan reply))
+  syncGroups := make(map[string]*sync.WaitGroup)
+
+  for _, rpc := range []string{NodeMetadataRPC, LastCrossLinkRPC} {
+    replyChannels[rpc] = make(chan reply, len(shardMap))
+    switch rpc {
+    case NodeMetadataRPC:
+      var mGroup sync.WaitGroup
+      syncGroups[rpc] = &mGroup
+    case LastCrossLinkRPC:
+      var lGroup sync.WaitGroup
+      syncGroups[rpc] = &lGroup
+    }
+  }
+
+  for i := 0; i < poolSize; i++ {
+    go m.worker(jobs, replyChannels, syncGroups)
+  }
+
+  type r struct {
+		Result NodeMetadataReply `json:"result"`
+	}
+
+  type processedCrossLink struct {
+    BlockNum    int
+    CrossLink   CrossLink
+    TS          time.Time
+  }
+
+  lastProcessed := make(map[int]processedCrossLink)
+  for now := range time.Tick(time.Duration(interval) * time.Second) {
+		queryID := 0
+		// Send requests to find potential shard 0 leaders
+		for n := range shardMap {
+      if shardMap[n] == 0 {
+			   nodeRequestFields["id"] = strconv.Itoa(queryID)
+			   requestBody, _ := json.Marshal(nodeRequestFields)
+			   jobs <- work{n, NodeMetadataRPC, requestBody}
+			   queryID++
+			   syncGroups[NodeMetadataRPC].Add(1)
+      }
+		}
+		syncGroups[NodeMetadataRPC].Wait()
+		close(replyChannels[NodeMetadataRPC])
+
+		leader := []string{}
+		for d := range replyChannels[NodeMetadataRPC] {
+			if d.oops == nil {
+				oneReport := r{}
+				json.Unmarshal(d.rpcResult, &oneReport)
+				if oneReport.Result.IsLeader {
+					leader = append(leader, d.address)
+				}
+			}
+		}
+
+    queryID = 0
+    // Request from all potential leaders
+    for _, l := range leader {
+      crossLinkRequestFields["id"] = strconv.Itoa(queryID)
+      requestBody, _ := json.Marshal(crossLinkRequestFields)
+      jobs <- work{l, LastCrossLinkRPC, requestBody}
+      queryID++
+      syncGroups[LastCrossLinkRPC].Add(1)
+    }
+    syncGroups[LastCrossLinkRPC].Wait()
+		close(replyChannels[LastCrossLinkRPC])
```

Comment:
please come around and use gofmt as a plugin in your editor. have it run on file save

Reply:
Forgot to run gofmt on the file.